src/ripple/app/tx/impl/SetSignerList.cpp

```diff
@@ -78,31 +88,77 @@ SetSignerList::determineOperation(
 NotTEC
 SetSignerList::preflight(PreflightContext const& ctx)
 {
-    auto const ret = preflight1(ctx);
-    if (!isTesSuccess(ret))
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
         return ret;
 
-    auto const result = determineOperation(ctx.tx, ctx.flags, ctx.j);
-    if (std::get<0>(result) != tesSUCCESS)
-        return std::get<0>(result);
-
-    if (std::get<3>(result) == unknown)
+    if (ctx.rules.enabled(featureSimplifiedSetSignerList) &&
+        ctx.rules.enabled(featureMultiSignReserve))
     {
-        // Neither a set nor a destroy.  Malformed.
-        JLOG(ctx.j.trace())
-            << "Malformed transaction: Invalid signer set list format.";
-        return temMALFORMED;
-    }
+        auto const quorum = ctx.tx[sfSignerQuorum];
+
+        // The quorum says when we are adding or removing the signer list:
+        if (quorum != 0)
+        {
+            auto const sl = deserializeSignerList(ctx.tx);
+
+            if (sl.empty())
+                return temMALFORMED;
+
+            if (sl.size() > STTx::maxMultiSigners)
+                return temMALFORMED;
+
+            if (!std::is_sorted(sl.begin(), sl.end()))
+                return temMALFORMED;
+
+            if (auto const dup = std::adjacent_find(sl.begin(), sl.end());
+                dup != sl.end())
+                return temBAD_SIGNER;
 
-    if (std::get<3>(result) == set)
+            auto const self = ctx.tx.getAccountID(sfAccount);
+            std::uint32_t weight = 0;
+
+            for (auto const& s : sl)
+            {
+                if (s.first == self)
+                    return temBAD_SIGNER;
+
+                if (s.second == 0)
+                    return temBAD_WEIGHT;
+
+                weight += s.second;
+            }
+
+            if (weight < quorum)
+                return temBAD_QUORUM;
+        }
+
+        if (quorum == 0 && ctx.tx.isFieldPresent(sfSignerEntries))
+            return temMALFORMED;
+    }
+    else
     {
-        // Validate our settings.
-        auto const account = ctx.tx.getAccountID(sfAccount);
-        NotTEC const ter = validateQuorumAndSignerEntries(
-            std::get<1>(result), std::get<2>(result), account, ctx.j);
-        if (ter != tesSUCCESS)
+        auto const result = determineOperation(ctx.tx);
+        if (std::get<0>(result) != tesSUCCESS)
+            return std::get<0>(result);
+
+        if (std::get<3>(result) == unknown)
```

Comment:
How about creating some named constants for these fields of the tuple, e.g. `TEC_` and `OPERATION_`?